### PR TITLE
[Feature] Add DIC

### PR DESCRIPTION
**Pull request by Yshuo-Li**

### CODE REVIEW

File: mmedit/models/restorers/dic.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        # generate
+        sr_list, heatmap_list = self.generator.forward(lq)
+
+        # loss
+        losses = OrderedDict()
+
+        loss_pix = 0.0
+        loss_align = 0.0
+        for step, (sr, heatmap) in enumerate(zip(sr_list, heatmap_list)):
+            losses[f'loss_pixel_v{step}'] = self.pixel_loss(sr, gt)
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/dic.py:101]: Here what do `v1, v2, ...` mean? Different components?

File: mmedit/models/restorers/dic.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            results = dict(lq=lq.cpu(), output=pred.cpu())
+            if gt is not None:
+                results['gt'] = gt.cpu()
+
+        # save image
+        if save_image:
+            if 'gt_path' in meta[0]:
+                the_path = meta[0]['gt_path']
+            else:
+                the_path = meta[0]['lq_path']
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/dic.py:172]: What does `the_path` mean? Why use `the`?

File: tests/test_dic_model.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import numpy as np
+import pytest
+import torch
+from mmcv.runner import obj_from_dict
+from mmcv.utils.config import Config
+
+from mmedit.models.builder import build_model
+
+
+def test_dic_midel():
```
Comment (ckkelvinchan) [inline tests/test_dic_model.py:10]: midel -> model

File: mmedit/models/restorers/dic.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            elif iteration is None:
+                save_path = osp.join(save_path, f'{folder_name}.png')
+            else:
+                raise ValueError('iteration should be number or None, '
+                                 f'but got {type(iteration)}')
+            mmcv.imwrite(tensor2img(pred), save_path)
+
+        return results
+
+    def val_step(self, data_batch, **kwargs):
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/dic.py:186]: This can be inherited from `BasicRestorer`.

File: mmedit/models/restorers/dic.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            lq (Tensor): Input lq images.
+            gt (Tensor): Ground-truth image. Default: None.
+            test_mode (bool): Whether in test mode or not. Default: False.
+            kwargs (dict): Other arguments.
+        """
+
+        if test_mode:
+            return self.forward_test(lq, gt=gt, **kwargs)
+
+        return self.generator.forward(lq)
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/dic.py:72]: It seems that this line will not be executed anyway?

Comment (ckkelvinchan) [reply mmedit/models/restorers/dic.py:72]: If line 93 uses `self(**data_batch, test_mode=False)`, this function can be inherited from `BasicRestorer`.

File: mmedit/models/restorers/dic.py (hunk 1 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/dic.py:101]: different steps

File: mmedit/models/restorers/dic.py (hunk 2 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/dic.py:172]: get `pred_path` from `gt_path` or `lq_path`

File: tests/test_dic_model.py (hunk 3 above)
Comment (Yshuo-Li) [reply tests/test_dic_model.py:10]: done

File: mmedit/models/restorers/dic.py (hunk 5 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/dic.py:72]: This is not a good suggestion.
`train_step` in `BasicRestorer` includes loss and optimizer.

File: mmedit/models/restorers/dic.py (hunk 2 above)
Comment (ckkelvinchan) [reply mmedit/models/restorers/dic.py:172]: How about directly using `pred_path`?

**Review by ckkelvinchan — COMMENTED** [review]
One comment about `the_path`. And there are many lines not covered by unittest.

File: mmedit/models/restorers/dic.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                 test_cfg=None,
+                 pretrained=None):
+        super(BasicRestorer, self).__init__()
+
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+        # model
+        self.generator = build_backbone(generator)
+        self.img_denormalize = ImgNormalize(
```
Comment (innerlee) [inline mmedit/models/restorers/dic.py:39]: Where is it used?

Comment (innerlee) [reply mmedit/models/restorers/dic.py:39]: Out of curiosity, where is the normalize corresponds to this denormalize?

Comment (Yshuo-Li) [reply mmedit/models/restorers/dic.py:39]: The normalize is in pipeline (called in the config file, similar to other method of this repo).

Comment (innerlee) [reply mmedit/models/restorers/dic.py:39]: How about the first question

Comment (Yshuo-Li) [reply mmedit/models/restorers/dic.py:45]: Q: Where is it used?
A: Deleted by mistake in the previous version (f0b85d2).

This problem has been fixed.

**Review by innerlee — APPROVED** [review]

File: mmedit/models/restorers/dic.py (hunk 6 above)
Comment (innerlee) [reply mmedit/models/restorers/dic.py:45]: Can this de-normalization be configurable? Just asking... We can leave the better normalize-denormalize story to future work. Currently they are quite ad-hoc.

File: tests/test_dic_model.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        assert result['eval_result'].keys() == set({'PSNR', 'SSIM'})
+        assert isinstance(result['eval_result']['PSNR'], np.float64)
+        assert isinstance(result['eval_result']['SSIM'], np.float64)
+
+        with pytest.raises(AssertionError):
+            # evaluation with metrics must have gt images
+            restorer(lq=inputs.cuda(), test_mode=True)
+
+
+if __name__ == '__main__':
```
Comment (innerlee) [inline tests/test_dic_model.py:98]: remove this from tests

File: mmedit/models/restorers/dic.py (hunk 6 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/dic.py:45]: post-pipeline?

Comment (innerlee) [reply mmedit/models/restorers/dic.py:45]: Maybe...
